sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/cache/LoadingWatermarkCache.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.gcp.spanner.changestreams.cache;
+
+import com.google.cloud.Timestamp;
+import java.time.Duration;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.io.gcp.spanner.changestreams.dao.PartitionMetadataDao;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.cache.CacheBuilder;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.cache.CacheLoader;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.cache.LoadingCache;
+
+public class LoadingWatermarkCache implements WatermarkCache {
+  private static final Object MIN_WATERMARK_KEY = new Object();
+
+  private final LoadingCache<Object, Optional<Timestamp>> cache;
+
+  public LoadingWatermarkCache(PartitionMetadataDao dao, org.joda.time.Duration refreshRate) {
+    this.cache =
+        CacheBuilder.newBuilder()
+            .refreshAfterWrite(Duration.ofMillis(refreshRate.getMillis()))
```

Comment:
think you need to make the loading async

I think you can use this method:
https://guava.dev/releases/22.0/api/docs/com/google/common/cache/CacheLoader.html#asyncReloading-com.google.common.cache.CacheLoader-java.util.concurrent.Executor-

Reply:
Done